### PR TITLE
data bytes or json from bytes

### PR DESCRIPTION
**Pull request by DimaCantemir**
added ability to create DataBytesOrJSON struct directly from base64 bytes. Useful when you receive an account though alternative channels, like Geyser plugin + graph protocol,  and need to reconstruct it with as little overhead as possible

### CODE REVIEW

File: rpc/types.go [hunk 1]
```diff
@@ -255,11 +255,15 @@ func DataBytesOrJSONFromBase64(stringBase64 string) (*DataBytesOrJSON, error) {
 	if err != nil {
 		return nil, err
 	}
+	return DataBytesOrJSONFromBase64Bytes(decoded)
+}
+
+func DataBytesOrJSONFromBase64Bytes(data []byte) (*DataBytesOrJSON, error) {
```
Comment (gagliardetto) [inline rpc/types.go:261]:
```suggestion

// DataBytesOrJSONFromBase64Bytes creates a new `DataBytesOrJSON` from the provided base64 bytes.
func DataBytesOrJSONFromBase64Bytes(b64Data []byte) (*DataBytesOrJSON, error) {
```

**Comment by aspin** [comment]
could we get a bump on this @gagliardetto ?

File: rpc/types.go [hunk 2]
```diff
@@ -255,11 +255,16 @@ func DataBytesOrJSONFromBase64(stringBase64 string) (*DataBytesOrJSON, error) {
...
+}
+
+// DataBytesOrJSONFromBase64Bytes creates a new `DataBytesOrJSON` from the provided base64 bytes.
+func DataBytesOrJSONFromBase64Bytes(b64Data []byte) (*DataBytesOrJSON, error) {
 	return &DataBytesOrJSON{
 		rawDataEncoding: solana.EncodingBase64,
 		asDecodedBinary: solana.Data{
 			Encoding: solana.EncodingBase64,
-			Content:  decoded,
+			Content:  b64Data,
```
Comment (gagliardetto) [inline rpc/types.go:267]:
```suggestion
			Content:  content,
```

**Comment by gagliardetto** [comment]
In that case, your PR content won't solve your problem.

Can't you just `DataBytesOrJSONFromBase64(string(encodedData))` ?

**Comment by gagliardetto** [comment]
In that case you have the **A** case, and need to create a `DataBytesOrJSON`. I see.

File: rpc/types.go [hunk 3]
```diff
@@ -255,11 +255,16 @@ func DataBytesOrJSONFromBase64(stringBase64 string) (*DataBytesOrJSON, error) {
...
 		},
 	}, nil
```
Comment (gagliardetto) [inline rpc/types.go:269]:
```suggestion
	}
```

File: rpc/types.go [hunk 4]
```diff
@@ -251,17 +251,22 @@ type DataBytesOrJSON struct {
...
-	decoded, err := base64.StdEncoding.DecodeString(stringBase64)
+	b64Data, err := base64.StdEncoding.DecodeString(stringBase64)
 	if err != nil {
 		return nil, err
 	}
+	return DataBytesOrJSONFromBytes(b64Data), nil
+}
+
+// DataBytesOrJSONFromBytes creates a new `DataBytesOrJSON` from the provided base64 bytes.
+func DataBytesOrJSONFromBytes(b64Data []byte) *DataBytesOrJSON {
```
Comment (gagliardetto) [inline rpc/types.go:262]:
```suggestion
// DataBytesOrJSONFromBytes creates a new `DataBytesOrJSON` from the provided bytes.
func DataBytesOrJSONFromBytes(data []byte) *DataBytesOrJSON {
```

File: rpc/types.go [hunk 5]
```diff
@@ -251,17 +251,22 @@ type DataBytesOrJSON struct {
...
 	return &DataBytesOrJSON{
 		rawDataEncoding: solana.EncodingBase64,
 		asDecodedBinary: solana.Data{
 			Encoding: solana.EncodingBase64,
-			Content:  decoded,
+			Content:  b64Data,
```
Comment (gagliardetto) [inline rpc/types.go:267]:
```suggestion
			Content:  data,
```

**Comment by gagliardetto** [comment]
LGTM, just need to address the couple above naming changes.

**Comment by DimaCantemir** [comment]
Done

File: rpc/types.go [hunk 6]
```diff
@@ -251,17 +251,22 @@ type DataBytesOrJSON struct {
...
 func DataBytesOrJSONFromBase64(stringBase64 string) (*DataBytesOrJSON, error) {
-	decoded, err := base64.StdEncoding.DecodeString(stringBase64)
+	b64Data, err := base64.StdEncoding.DecodeString(stringBase64)
 	if err != nil {
 		return nil, err
 	}
+	return DataBytesOrJSONFromBytes(b64Data), nil
+}
+
+// DataBytesOrJSONFromBytes creates a new `DataBytesOrJSON` from the provided base64 bytes.
```
Comment (gagliardetto) [inline rpc/types.go:261]:
```suggestion
// DataBytesOrJSONFromBytes creates a new `DataBytesOrJSON` from the provided bytes.
```

File: rpc/types.go [hunk 7]
```diff
@@ -251,17 +251,22 @@ type DataBytesOrJSON struct {
 }
 
 func DataBytesOrJSONFromBase64(stringBase64 string) (*DataBytesOrJSON, error) {
-	decoded, err := base64.StdEncoding.DecodeString(stringBase64)
+	b64Data, err := base64.StdEncoding.DecodeString(stringBase64)
 	if err != nil {
 		return nil, err
 	}
+	return DataBytesOrJSONFromBytes(b64Data), nil
```
Comment (gagliardetto) [inline rpc/types.go:258]:
```suggestion
	decodedData, err := base64.StdEncoding.DecodeString(stringBase64)
	if err != nil {
		return nil, err
	}
	return DataBytesOrJSONFromBytes(decodedData), nil
```

**Review by gagliardetto — CHANGES_REQUESTED** [review]

**Comment by gagliardetto** [comment]
Still missing the change in the comment.

If you go to the "Files changed" tab, you can commit suggestions directly.

**Comment by DimaCantemir** [comment]
Done. When you merge this could you please tag a new release ? our devOps people are itching to run some performance tests after we integrate this change

**Comment by gagliardetto** [comment]
No release is scheduled for today.

You can use the latest source with your changes by `go get github.com/gagliardetto/solana-go@8bd6977469ecb5bbd8a27488648b82bd1904f5fe`

**Pull request merged**